File: jiuwenswarm/channels/ide/packages/vscode-extension/src/client/SessionManager.ts
import { WsClient } from './WsClient';
import { JiuwenMessage, SessionInfo, makeRequest } from './protocol';

type PendingRequest = {
  resolve: (payload: Record<string, unknown>) => void;
  reject: (err: Error) => void;
  timer: NodeJS.Timeout;
};

const REQUEST_TIMEOUT_SEC = 15;

export class SessionManager {
  private _sessionId: string | null = null;
  private _sessionTitle = 'JiuwenSwarm';
  private pending = new Map<string, PendingRequest>();
  private channelId: string;
  private sessionListeners: ((sid: string | null) => void)[] = [];

  constructor(private readonly ws: WsClient, channelId = 'ide') {
    this.channelId = channelId;
    this.ws.on('message', (msg) => this.handleMessage(msg));
  }

  get sessionId(): string | null { return this._sessionId; }
  get sessionTitle(): string { return this._sessionTitle; }

  addSessionListener(l: (sid: string | null) => void): void {
    this.sessionListeners.push(l);
  }

  removeSessionListener(l: (sid: string | null) => void): void {
    const idx = this.sessionListeners.indexOf(l);
    if (idx >= 0) this.sessionListeners.splice(idx, 1);
  }

  /** Returns available models and the currently active model name. */
  async listModels(): Promise<{ models: Record<string, unknown>[]; activeModel: string | null }> {
    const payload = await this.request('models.list', {});
    const models = (payload.models as Record<string, unknown>[]) || [];
    const activeModel = payload.active_model as string | undefined || null;
    return { models, activeModel };
  }

  /** List recent sessions */
  async listSessions(limit = 20): Promise<SessionInfo[]> {
    const payload = await this.request('session.list', { limit });
    const arr = payload.sessions as SessionInfo[] | undefined;
    if (!arr) return [];
    return arr.map((obj) => ({
      session_id: (obj.session_id as string) || '',
      title: obj.title as string | undefined,
      last_message_at: obj.last_message_at as number | undefined,
      created_at: obj.created_at as number | undefined,
      message_count: obj.message_count as number | undefined,
      history_path: (obj.history_path as string) || '',
    }));
  }

  /** Switch to an existing session */
  async switchSession(sid: string, mode?: string): Promise<void> {
    const params: Record<string, unknown> = { session_id: sid };
    if (mode) params.mode = mode;
    const payload = await this.request('session.switch', params, sid);
    this.setSessionId(sid);
    this._sessionTitle = (payload.title as string) || sid;
  }

  /**
   * Fire-and-forget chat message. Streaming events arrive via WsClient 'message' events.
   * If [ideContext] is non-null it is prepended to [content].
   * If [mediaItems] is non-null, image attachments are included in base64 format.
   */
  sendChat(
    content: string,
    mode: string,
    requestId: string,
    ideContext?: string,
    mediaItems?: unknown[],
    modelName?: string,
  ): boolean {
    const sid = this._sessionId;
    if (!sid) return false;
    const fullContent = ideContext && ideContext.trim()
      ? `${content}\n\n${ideContext}`
      : content;
    const params: Record<string, unknown> = {
      content: fullContent,
      mode,
      session_id: sid,
    };
    if (modelName) {
      params.model_name = modelName;
    }
    if (mediaItems && mediaItems.length > 0) {
      params.media_items = mediaItems;
    }
    const msg = makeRequest('chat.send', params, this.channelId);
    msg.id = requestId;
    return this.ws.send(msg);
  }

  /**
   * Fire-and-forget interrupt-resume answer (e.g. plan approval for exit_plan_mode).
   * [requestId] is the id of the original chat.send that triggered the interrupt; the
   * server resumes that suspended turn. We reuse it as msg.id so the resumed stream
   * keeps mapping to the same pending turn in the webview.
   */
  sendAnswer(
    requestId: string,
    answers: unknown[],
    source: string,
    mode: string,
  ): boolean {
    const sid = this._sessionId;
    if (!sid) return false;
    const msg = makeRequest('chat.send', {
      session_id: sid,
      request_id: requestId,
      answers,
      source,
      mode,
    }, this.channelId);
    msg.id = requestId;
    return this.ws.send(msg);
  }

  /** Fetch current process & system memory usage (MB). */
  async getMemoryUsage(): Promise<{ rss_mb: number; total_mb: number; available_mb: number }> {
    const payload = await this.request('memory.compute', {});
    return {
      rss_mb: (payload.rss_mb as number) || 0,
      total_mb: (payload.total_mb as number) || 0,
      available_mb: (payload.available_mb as number) || 0,
    };
  }

  /** List available skills */
  async listSkills(): Promise<Record<string, unknown>[]> {
    const payload = await this.request('skills.list', {});
    return (payload.skills as Record<string, unknown>[]) || [];
  }

  /** Enable/disable a skill */
  async toggleSkill(skillId: string, enabled: boolean): Promise<void> {
    await this.request('skills.toggle', { skill_id: skillId, enabled });
  }

  /** Delete a session */
  async deleteSession(sid: string): Promise<void> {
    await this.request('session.delete', { session_id: sid }, sid);
  }

  /** Rename the active session (empty title clears it). */
  async renameSession(title: string): Promise<void> {
    const sid = this._sessionId;
    if (!sid) return;
    await this.request('session.rename', { session_id: sid, title });
    this._sessionTitle = title.trim() || sid;
  }

  /**
   * Fire-and-forget request to load paginated session history.
   * The server streams back history.message events through the WebSocket which
   * flow to all message listeners (including ChatPanel.onJiuwenMessage).
   * A history.done event marks the end of the page.
   */
  loadHistory(sid: string, pageIdx = 1): boolean {
    const msg = makeRequest('history.get', { session_id: sid, page_idx: pageIdx }, this.channelId);
    return this.ws.send(msg);
  }

  // ──────────────────────────────────────────
  // Internal
  // ──────────────────────────────────────────

  private request(method: string, params: Record<string, unknown>, sid?: string): Promise<Record<string, unknown>> {
    const p = { ...params };
    if (sid) p.session_id = sid;
    const msg = makeRequest(method, p, this.channelId);
    return new Promise((resolve, reject) => {
      const id = msg.id!;
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new Error(`Request '${method}' timed out`));
      }, REQUEST_TIMEOUT_SEC * 1000);

      this.pending.set(id, { resolve, reject, timer });
      if (!this.ws.send(msg)) {
        clearTimeout(timer);
        this.pending.delete(id);
        reject(new Error('WebSocket not connected'));
      }
    });
  }

  private handleMessage(msg: JiuwenMessage): void {
    // ── E2A format responses ──
    const responseKind = msg.response_kind
      || (msg.payload?.response_kind as string | undefined);
    if (responseKind) {
      const rid = msg.request_id
        || (msg.payload?.request_id as string | undefined);
      if (!rid) return;
      const future = this.pending.get(rid);
      if (!future) return;
      if (responseKind === 'e2a.complete') {
        const body = (msg.body as Record<string, unknown>)
          || (msg.payload?.body as Record<string, unknown>)
          || {};
        const result = (body.result as Record<string, unknown>) || body;
        clearTimeout(future.timer);
        this.pending.delete(rid);
        future.resolve(result);
      } else if (responseKind === 'e2a.error') {
        const body = (msg.body as Record<string, unknown>)
          || (msg.payload?.body as Record<string, unknown>)
          || {};
        const err = (body.message as string) || 'E2A error';
        clearTimeout(future.timer);
        this.pending.delete(rid);
        future.reject(new Error(err));
      } else {
        clearTimeout(future.timer);
        this.pending.delete(rid);
        future.reject(new Error(`Unexpected response_kind: ${responseKind}`));
      }
      return;
    }

    // ── Gateway legacy format responses ──
    if (msg.type === 'res' && msg.id) {
      const future = this.pending.get(msg.id);
      if (future) {
        clearTimeout(future.timer);
        this.pending.delete(msg.id);
        if (msg.ok) {
          future.resolve((msg.payload as Record<string, unknown>) || {});
        } else {
          const err =
            (msg.payload as Record<string, unknown> | undefined)?.error as string | undefined
            || 'Request failed';
          future.reject(new Error(err));
        }
      }
      return;
    }

    // ── Events (connection.ack gives us the session) ──
    if (msg.type === 'event' && msg.event) {
      const payload = (msg.payload as Record<string, unknown>) || {};
      if (msg.event === 'connection.ack') {
        const sid = payload.session_id as string | undefined;
        if (sid) {
          this.setSessionId(sid);
        }
      }
    }
  }

  private setSessionId(sid: string | null): void {
    if (this._sessionId === sid) return;
    this._sessionId = sid;
    this.sessionListeners.forEach((l) => l(sid));
  }
}

File: jiuwenswarm/channels/ide/packages/jetbrains-plugin/src/main/kotlin/com/jiuwenswarm/plugin/client/SessionManager.kt
package com.jiuwenswarm.plugin.client

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.logger
import java.util.UUID
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.TimeUnit

private val LOG = logger<SessionManager>()
private val gson = Gson()
private const val REQUEST_TIMEOUT_SEC = 5L

typealias SessionChangeListener = (String?) -> Unit

data class SessionInfo(
    val session_id: String,
    val title: String?,
    val last_message_at: Long?,
    val created_at: Long?,
    val message_count: Int?,
    val history_path: String? = null,
)

class SessionManager(
    private val ws: WsClient,
    private val channelId: String = "ide",
) {
    var sessionId: String? = null
        private set
    var sessionTitle: String = "JiuwenSwarm"
        private set

    private val pending = ConcurrentHashMap<String, CompletableFuture<JsonObject>>()
    private val sessionListeners = CopyOnWriteArrayList<SessionChangeListener>()
    private val listener: MessageListener = { msg -> handleMessage(msg) }

    init {
        ws.addMessageListener(listener)
    }

    fun addSessionListener(l: SessionChangeListener) = sessionListeners.add(l)
    fun removeSessionListener(l: SessionChangeListener) = sessionListeners.remove(l)

    fun dispose() {
        ws.removeMessageListener(listener)
        pending.values.forEach { it.cancel(true) }
        pending.clear()
    }

    fun listSessions(limit: Int = 20): List<SessionInfo> {
        val payload = request("session.list", mapOf("limit" to limit))
        val arr = payload.getAsJsonArray("sessions") ?: return emptyList()
        return arr.map { el ->
            val obj = el.asJsonObject
            SessionInfo(
                session_id = obj.get("session_id")?.asString ?: "",
                title = obj.get("title")?.asString,
                last_message_at = obj.get("last_message_at")?.asLong,
                created_at = obj.get("created_at")?.asLong,
                message_count = obj.get("message_count")?.asInt,
                history_path = obj.get("history_path")?.asString ?: "",
            )
        }
    }

    /** Returns available models and the currently active model name. */
    fun listModels(): Pair<List<JsonObject>, String?> {
        val payload = request("models.list", emptyMap())
        val models = payload.getAsJsonArray("models")?.map { it.asJsonObject } ?: emptyList()
        val activeModel = payload.get("active_model")?.asString
        return models to activeModel
    }

    /**
     * Fetches current process and system memory usage (MB) from the server.
     * Returns a triple (rss_mb, total_mb, available_mb).
     */
    fun getMemoryUsage(): Triple<Double, Double, Double> {
        val payload = request("memory.compute", emptyMap())
        val rss = payload.get("rss_mb")?.asDouble ?: 0.0
        val total = payload.get("total_mb")?.asDouble ?: 0.0
        val available = payload.get("available_mb")?.asDouble ?: 0.0
        return Triple(rss, total, available)
    }

    /**
     * Returns the list of available skills registered with the jiuwenswarm instance.
     * Each [JsonObject] contains at minimum "skill_id" and "name".
     * Throws if the server does not support the skills.list method.
     */
    fun listSkills(): List<JsonObject> {
        val payload = request("skills.list", emptyMap())
        return payload.getAsJsonArray("skills")?.map { it.asJsonObject } ?: emptyList()
    }

    /**
     * Enables or disables a skill by its ID.
     * Throws if the server does not support skills.toggle.
     */
    fun toggleSkill(skillId: String, enabled: Boolean) {
        request("skills.toggle", mapOf("skill_id" to skillId, "enabled" to enabled))
    }

    /**
     * Permanently deletes a session by its ID.
     * Throws if the server does not support session.delete or the session does not exist.
     */
    fun deleteSession(sid: String) {
        request("session.delete", mapOf("session_id" to sid))
    }

    /** Rename the active session (empty title clears it). Updates the cached title. */
    fun renameSession(title: String) {
        val sid = sessionId ?: return
        val payload = request("session.rename", mapOf("title" to title), sid)
        sessionTitle = payload.get("title")?.asString ?: title
    }

    fun switchSession(sid: String, mode: String? = null) {
        val params = mutableMapOf("session_id" to sid)
        if (mode != null) params["mode"] = mode
        val payload = request("session.switch", params, sid)
        setSessionId(sid)
        sessionTitle = payload.get("title")?.asString ?: sid
    }

    /**
     * Fire-and-forget request to load paginated session history.
     * The server streams back [history.message] events through the WebSocket which
     * flow to all registered message listeners (including [ChatToolWindow.onJiuwenMessage]).
     * A [history.done] event marks the end of the page.
     */
    fun loadHistory(sid: String, pageIdx: Int = 1): Boolean {
        val id = UUID.randomUUID().toString()
        val msg = buildJsonObject {
            addProperty("id", id)
            addProperty("type", "req")
            addProperty("channel_id", channelId)
            addProperty("method", "history.get")
            add("params", gson.toJsonTree(
                mapOf("session_id" to sid, "page_idx" to pageIdx)
            ).asJsonObject)
            addProperty("timestamp", System.currentTimeMillis() / 1000.0)
        }
        return ws.send(msg)
    }

    /**
     * Fire-and-forget chat message. Streaming events arrive via WsClient message listener.
     * If [ideContext] is non-null it is prepended to [content] as a structured block so the
     * agent can see the current file, selection, and diagnostics without the user having to
     * copy-paste them manually.
     * If [mediaItems] is non-null, image attachments are included in base64 format.
     */
    fun sendChat(
        content: String,
        mode: String,
        requestId: String,
        ideContext: String? = null,
        mediaItems: com.google.gson.JsonArray? = null,
        model: String? = null,
    ): Boolean {
        val sid = sessionId ?: return false
        val fullContent = if (!ideContext.isNullOrBlank()) {
            "$content\n\n$ideContext"
        } else {
            content
        }
        val msg = buildJsonObject {
            addProperty("id", requestId)
            addProperty("type", "req")
            addProperty("channel_id", channelId)
            addProperty("method", "chat.send")
            add("params", buildJsonObject {
                addProperty("content", fullContent)
                addProperty("mode", mode)
                addProperty("session_id", sid)
                if (model != null) {
                    addProperty("model_name", model)
                }
                if (mediaItems != null && mediaItems.size() > 0) {
                    add("media_items", mediaItems)
                }
            })
        }
        return ws.send(msg)
    }

    /**
     * Fire-and-forget interrupt-resume answer (e.g. plan approval for exit_plan_mode).
     * [requestId] is the id of the original chat.send that triggered the interrupt; the
     * server resumes that suspended turn. We reuse it as msg.id so the resumed stream
     * keeps mapping to the same pending turn in the webview.
     */
    fun sendAnswer(
        requestId: String,
        answers: com.google.gson.JsonArray,
        source: String,
        mode: String,
    ): Boolean {
        val sid = sessionId ?: return false
        val msg = buildJsonObject {
            addProperty("id", requestId)
            addProperty("type", "req")
            addProperty("channel_id", channelId)
            addProperty("method", "chat.send")
            add("params", buildJsonObject {
                addProperty("session_id", sid)
                addProperty("request_id", requestId)
                add("answers", answers)
                addProperty("source", source)
                addProperty("mode", mode)
            })
        }
        return ws.send(msg)
    }

    fun interrupt() {
        val msg = buildJsonObject {
            addProperty("id", java.util.UUID.randomUUID().toString())
            addProperty("type", "req")
            addProperty("channel_id", channelId)
            addProperty("method", "chat.interrupt")
            add("params", buildJsonObject { })
        }
        ws.send(msg)
    }

    // ──────────────────────────────────────────
    private fun setSessionId(sid: String?) {
        if (sessionId == sid) return
        sessionId = sid
        sessionListeners.forEach { it(sid) }
    }

    private fun request(
        method: String,
        params: Map<String, Any>,
        sid: String? = null,
    ): JsonObject {
        // Blocking wait (up to REQUEST_TIMEOUT_SEC) must never happen on the EDT or the UI freezes.
        // All call sites dispatch to a pooled/background thread first; fail fast if a future caller does not.
        check(!ApplicationManager.getApplication().isDispatchThread) {
            "SessionManager.request('$method') must not be called from the EDT"
        }
        val id = UUID.randomUUID().toString()
        val paramsWithSession = params.toMutableMap()
        if (sid != null) paramsWithSession["session_id"] = sid
        val msg = buildJsonObject {
            addProperty("id", id)
            addProperty("type", "req")
            addProperty("channel_id", channelId)
            addProperty("method", method)
            add("params", gson.toJsonTree(paramsWithSession).asJsonObject)
            addProperty("timestamp", System.currentTimeMillis() / 1000.0)
        }
        val future = CompletableFuture<JsonObject>()
        pending[id] = future
        if (!ws.send(msg)) {
            pending.remove(id)
            throw IllegalStateException("WebSocket not connected")
        }
        return try {
            future.get(REQUEST_TIMEOUT_SEC, TimeUnit.SECONDS)
        } catch (e: Exception) {
            pending.remove(id)
            throw RuntimeException("Request '$method' failed: ${e.javaClass.simpleName} - ${e.message}", e)
        }
    }

    private fun handleMessage(msg: JsonObject) {
        val msgType = msg.get("type")?.asString

        // ── E2A format responses (gateway may wrap session.list/switch in E2A) ──
        val responseKind = msg.get("response_kind")?.asString
        if (responseKind != null) {
            val rid = msg.get("request_id")?.asString ?: return
            val future = pending.remove(rid) ?: return   // not our request — ignore
            when (responseKind) {
                "e2a.complete" -> {
                    val body = msg.getAsJsonObject("body") ?: JsonObject()
                    val result = body.getAsJsonObject("result") ?: body
                    future.complete(result)
                }
                "e2a.error" -> {
                    val body = msg.getAsJsonObject("body") ?: JsonObject()
                    val err = body.get("message")?.asString ?: "E2A error"
                    future.completeExceptionally(RuntimeException(err))
                }
                else -> future.completeExceptionally(RuntimeException("Unexpected response_kind: $responseKind"))
            }
            return
        }

        // ── Gateway legacy format responses ──
        if (msgType == "res") {
            val rid = msg.get("id")?.asString
            if (rid == null) {
                LOG.warn("Response missing 'id' field: ${msg}")
                return
            }
            val future = pending.remove(rid)
            if (future == null) {
                LOG.warn("Received response for unknown id=$rid; pending=${pending.keys}")
                return
            }
            if (msg.get("ok")?.asBoolean == true) {
                val payload = msg.getAsJsonObject("payload") ?: JsonObject()
                future.complete(payload)
            } else {
                val err = msg.get("error")?.asString
                    ?: msg.getAsJsonObject("payload")?.get("error")?.asString
                    ?: "Request failed"
                future.completeExceptionally(RuntimeException(err))
            }
            return
        }

        // ── Events (connection.ack gives us the session) ──
        if (msgType == "event") {
            val eventName = msg.get("event")?.asString ?: return
            val payload = msg.getAsJsonObject("payload") ?: JsonObject()
            when (eventName) {
                "connection.ack" -> {
                    val sid = payload.get("session_id")?.asString
                    if (sid != null) {
                        LOG.info("Auto-session from connection.ack: $sid")
                        setSessionId(sid)
                    }
                }
            }
            return
        }
    }
}

// ──────────────────────────────────────────────────────────────────
private fun buildJsonObject(block: JsonObject.() -> Unit): JsonObject =
    JsonObject().also(block)
